etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R

```diff
@@ -173,28 +210,159 @@ process_parcel_file <- function(s3_bucket_uri,
         filter(!is.na(lon) & !is.na(x_3435)) %>%
         bind_rows(spatial_df_missing)
     }
+    tictoc::toc()
 
-    # Sort by year, town code, and PIN10 for better compression
-    spatial_df_merged <- spatial_df_merged %>%
+    tictoc::tic(paste("Calculated shape features:", file_year))
+    # Simplify the planar geometry to reduce noise in shape-based features
+    # calculated below. Convert to a data.table of the vertices of each polygon.
+    # Using data.table here because it's much faster than dplyr
+    spatial_mat_coords <- spatial_df_merged %>%
+      st_set_geometry("geometry_3435") %>%
+      st_simplify(dTolerance = 2, preserveTopology = FALSE) %>%
+      st_cast("MULTIPOLYGON") %>%
+      st_coordinates() %>%
+      as.data.table()
+
+    # Use the coordinate vertices to calculate additional features about
+    # the shape of the polygon
+    spatial_mat_calc <- spatial_mat_coords[
+      ,
+      # Calculate the interior angle between each pair of points
+      angle := calculate_angles(as.matrix(.SD)),
+      .SDcols = c("X", "Y"),
+      by = c("L1", "L2", "L3")
```

Comment:
I'm having a hard time parsing what L1, L2, and L3 are?

Reply:
They are one of [the outputs](https://r-spatial.github.io/sf/reference/st_coordinates.html#value) from `st_coordinates()`. Basically, L3 corresponds to the PIN10 group while L1 and L2 can represent different sub-polygons of a multi-polygon.